cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+    {
+        const float multiplier = exaggeration * float(zNorm);
+        divergence             = 0.;
+
+        const IdxType prefetch_dist = 32;
+
+        daal::TlsSum<float, avx512> divTlsData(1);
+
+        SafeStatus safeStat;
+        daal::tls<float *> logTlsData([=, &safeStat]() {
+            auto logData = services::internal::service_scalable_calloc<float, avx512>(nElements);
+            if (!logData)
+            {
+                safeStat.add(services::ErrorMemoryAllocationFailed);
+            }
+            return logData;
+        });
+
+        const IdxType nThreads    = threader_get_threads_number();
+        const IdxType sizeOfBlock = services::internal::min<avx512, size_t>(256, N / nThreads + 1);
+        const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
+
+        daal::threader_for(nBlocks, nBlocks, [&](IdxType iBlock) {
+            const IdxType iStart = iBlock * sizeOfBlock;
+            const IdxType iEnd   = services::internal::min<avx512, IdxType>(N, iStart + sizeOfBlock);
+            float * logLocal     = logTlsData.local();
+            if (!logLocal) return;
+
+            float * divLocal = divTlsData.local();
+            DAAL_CHECK_MALLOC_THR(divLocal);
+
+            xyType<float> row_point;
+            IdxType iCol, prefetch_index;
+            float y1d, y2d, sqDist, PQ;
+
+            for (IdxType iRow = iStart; iRow < iEnd; ++iRow)
+            {
+                size_t iSize      = 0;
+                mem._attr[iRow].x = 0.0;
+                mem._attr[iRow].y = 0.0;
+                row_point         = mem._pos[iRow];
+
+                if (!DivComp)
+                {
+                    IdxType start_index = row[iRow] - 1;
+                    IdxType range       = row[iRow + 1] - row[iRow];
+
+                    __m512 vec_1   = _mm512_set1_ps(1.0);
+                    __m512i vec_1i = _mm512_set1_epi32(1);
+
+                    __m512 vec_point_x = _mm512_set1_ps(row_point.x);
+                    __m512 vec_point_y = _mm512_set1_ps(row_point.y);
+
+                    __m512 vec_point_xs = _mm512_setzero_ps();
+                    __m512 vec_point_ys = _mm512_setzero_ps();
+
+                    for (IdxType i = 0; i < (range / 16) * 16; i += 16)
+                    {
+                        prefetch_index = start_index + i + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
+
+                        __m512i vec_iCol = _mm512_sub_epi32(_mm512_loadu_epi32((__m512i *)&col[start_index + i]), vec_1i);
+
+                        __m512 vec_point_xd = _mm512_sub_ps(vec_point_x, _mm512_i32gather_ps(vec_iCol, &mem._pos[0].x, 8));
+                        __m512 vec_point_yd = _mm512_sub_ps(vec_point_y, _mm512_i32gather_ps(vec_iCol, &mem._pos[0].y, 8));
+
+                        __m512 vec_pq = _mm512_div_ps(
+                            _mm512_loadu_ps((__m512 *)&val[start_index + i]),
```

Comment:
Please don't use getting pointer to the value in array semantics. It can be dangerous since the last iteration can accidentally access elements outside the range and lead to segfault. 

```suggestion
                            _mm512_loadu_ps((__m512 *)(val + start_index + i)),
```